src/Assets/AssetDataRegistry.php

```diff
@@ -0,0 +1,246 @@
+				'weekdaysShort' => array_values( $wp_locale->weekday_abbrev ),
+			],
+			'orderStatuses' => $this->get_order_statuses( wc_get_order_statuses() ),
+			'siteTitle'     => get_bloginfo( 'name ' ),
+			'wcAssetUrl'    => plugins_url( 'assets/', WC_PLUGIN_FILE ),
+		];
+	}
+
+	/**
+	 * Returns block-related data for enqueued wc-settings script.
+	 * Format order statuses by removing a leading 'wc-' if present.
+	 *
+	 * @param array $statuses Order statuses.
+	 * @return array formatted statuses.
+	 */
+	protected function get_order_statuses( $statuses ) {
+		$formatted_statuses = array();
+		foreach ( $statuses as $key => $value ) {
+			$formatted_key                        = preg_replace( '/^wc-/', '', $key );
+			$formatted_statuses[ $formatted_key ] = $value;
+		}
+		return $formatted_statuses;
+	}
+
+	/**
+	 * Used for on demand initialization of asset data and registering it with
+	 * the internal data registry.
+	 *
+	 * Note: core data will overwrite any externally registered data via the api.
+	 */
+	protected function initialize_core_data() {
+		/**
+		 * Low level hook for registration of new data late in the cycle.
+		 *
+		 * Developers, do not use this hook as it is likely to be removed.
+		 * Instead, use the data api:
+		 * wc_blocks_container()
+		 *     ->get( Automattic\WooCommerce\Blocks\Assets\AssetDataRegistry::class )
+		 *     ->add( $key, $value )
+		 */
+		$settings = apply_filters(
+			'woocommerce_shared_settings',
+			$this->data
+		);
+		// note this WILL wipe any data already registered to these keys because
+		// they are protected.
+		$this->data = array_merge_recursive( $settings, $this->get_core_data() );
+	}
+
+	/**
+	 * Loops through each registered lazy data callback and adds the returned
+	 * value to the data array.
+	 *
+	 * This method is executed right before preparing the data for printing to
+	 * the rendered screen.
+	 *
+	 * @return void
+	 */
+	protected function execute_lazy_data() {
+		foreach ( $this->lazy_data as $key => $callback ) {
+			$this->data[ $key ] = $callback();
+		}
+	}
+
+	/**
+	 * Exposes private registered data to child classes.
+	 *
+	 * @return array  The registered data on the private data property
+	 */
+	protected function get() {
+		return $this->data;
+	}
+
+	/**
+	 * Interface for adding data to the registry.
+	 *
+	 * @param string $key  The key used to reference the data being registered.
+	 *                     You can only register data that is not already in the
+	 *                     registry identified by the given key.
+	 * @param mixed  $data If not a function, registered to the registry as is.
+	 *                     If a function, then the callback is invoked right
+	 *                     before output to the screen.
+	 *
+	 * @throws InvalidArgumentException  Only throws when site is in debug mode.
+	 */
+	public function add( $key, $data ) {
+		if ( ! is_string( $key ) ) {
+			if ( $this->debug() ) {
+				throw new InvalidArgumentException(
```

Comment:
Ya I wondered about that.  However, this is _explicitly_ an error intended for developers - meaning, it should surface really quickly if there's a problem.  I'm wary that doing_it_wrong notices/errors/logging would result in inaction which could lead to unexpected behaviour if ignored.

Reply:
`_doing_it_wrong`/`wc_doing_it_wrong` are also for developers. They only surface visually if debug notices are turned on and displayed, otherwise it's logged. I don't think this is much different to what you have, besides it being a notice vs an exception. In both cases, debug is required otherwise it's ignored.

Reply:
To clarify, when I wrote,

> I'm wary that doing_it_wrong notices/errors/logging would result in inaction which could lead to unexpected behaviour if ignored.

I mean that (from observation) due to how `doing_it_wrong` (wp core) has been used in various ways for everything from things that will "still work" but aren't correct implementation to things that wont' work - I decided that for here that it's important to actually break things (and throw an exception, because it's exceptional) - and fail catastrophically really early (assuming the developer is developing with debug mode on!).  It's less that the developer is "doing it wrong" and more that they have to choose a more unique key for the data property.

I think an improvement I can make here is to throw the exception always, but if in debug catch the exception and log it to `wc_caught_exception`.

Reply:
I implemented a slight improvement in 504c48176.  However, in another iteration after this is merged, I'd like to explore having the try/catch at the bootstrap level and it can handle any bubbled exceptions appropriately.  For instance it could handle when request type is api or in debug mode and log and prepare for user-friendly message (if necessary).

Reply:
Created https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/982 for this.